core/src/main/java/hudson/model/Queue.java

```diff
@@ -1630,6 +1632,11 @@ public void maintain() {
                     continue;
                 }
 
+                Label itemLabel = p.getAssignedLabel();
```

Comment:
One last thing, could you put this whole thing into the else clause. It doesn't affect flyweight tasks so its better to compartmentalize this code. You should also add a comment on exactly why we are doing this. The queue code is quite involved so its best to add it for a future reader.

Other than that. This change looks reasonable to me. Perhaps Vincent can chime in on what he thinks.